modules/markup/html.go

```diff
@@ -204,350 +141,522 @@ func IsSameDomain(s string) bool {
+// data, and ensures its validity. Transformations include: replacing links and
+// emails with HTML links, parsing shortlinks in the format of [[Link]], like
+// MediaWiki, linking issues in the format #ID, and mentions in the format
+// @user, and others.
+func PostProcess(
+	rawHTML []byte,
+	urlPrefix string,
+	metas map[string]string,
+	isWikiMarkdown bool,
+) ([]byte, error) {
+	// create the context from the parameters
+	ctx := &postProcessCtx{
+		metas:          metas,
+		urlPrefix:      urlPrefix,
+		isWikiMarkdown: isWikiMarkdown,
+		procs:          defaultProcessors,
+		visitLinksForShortLinks: true,
+	}
+	return ctx.postProcess(rawHTML)
+}
+
+var commitMessageProcessors = []processor{
+	mentionProcessor,
+	fullIssuePatternProcessor,
+	issueIndexPatternProcessor,
+	crossReferenceIssueIndexPatternProcessor,
+	fullSha1PatternProcessor,
+	sha1CurrentPatternProcessor,
+	emailAddressProcessor,
+	linkProcessor,
+}
+
+// RenderCommitMessage will use the same logic as PostProcess, but will disable
+// the shortLinkProcessor and will add a defaultLinkProcessor if defaultLink is
+// set, which changes every text node into a link to the passed default link.
+func RenderCommitMessage(
+	rawHTML []byte,
+	urlPrefix, defaultLink string,
+	metas map[string]string,
+) ([]byte, error) {
+	ctx := &postProcessCtx{
+		metas:     metas,
+		urlPrefix: urlPrefix,
+		procs:     commitMessageProcessors,
+	}
+	if defaultLink != "" {
+		// we don't have to fear data races, because being
+		// commitMessageProcessors of fixed len and cap, every time we append
+		// something to it the slice is realloc+copied, so append always
+		// generates the slice ex-novo.
+		ctx.procs = append(ctx.procs, genDefaultLinkProcessor(defaultLink))
+	}
+	return ctx.postProcess(rawHTML)
+}
+
+func (ctx *postProcessCtx) postProcess(rawHTML []byte) ([]byte, error) {
+	if ctx.procs == nil {
+		ctx.procs = defaultProcessors
+	}
+
+	// give a generous extra 50 bytes
+	res := make([]byte, 0, len(rawHTML)+50)
+	res = append(res, "<body>"...)
+	res = append(res, rawHTML...)
+	res = append(res, "</body>"...)
+
+	// parse the HTML
+	nodes, err := html.ParseFragment(bytes.NewReader(res), nil)
+	if err != nil {
+		return nil, &postProcessError{"invalid HTML", err}
+	}
+
+	for _, node := range nodes {
+		ctx.visitNode(node)
+	}
+
+	// Create buffer in which the data will be placed again. We know that the
+	// length will be at least that of res; to spare a few alloc+copy, we
+	// reuse res, resetting its length to 0.
+	buf := bytes.NewBuffer(res[:0])
+	// Render everything to buf.
+	for _, node := range nodes {
+		err = html.Render(buf, node)
+		if err != nil {
+			return nil, &postProcessError{"error rendering processed HTML", err}
 		}
-		sha := m[6]
-		var subtree string
-		if len(m) > 7 && len(m[7]) > 0 {
-			subtree = string(m[7])
+	}
+
+	// remove initial parts - because Render creates a whole HTML page.
+	const lenInit = len(`<html><head></head><body>`)
```

Comment:
Can this instead find `<body>` in string and respective closing tag to be more future proof?